### PR TITLE
[TRTLLM-12534][fix] Nemotron Nano - properly account for text prompts in inflight batching with EVS on

### PR DESCRIPTION
**Pull request by moraxu**
## Summary by CodeRabbit

* **Bug Fixes**
  * Improved multimodal video context handling in Nemotron Nano V2VL models.
  * Fixed token-positioning so mixed text-only and multimodal requests in a batch no longer overwrite each other.
  * Ensured multimodal embeddings are placed accurately within flattened token sequences.
  * Removed an overly-restrictive batch validation that rejected certain multimodal batching patterns.

* **Tests**
  * Added regression tests covering interspersed multimodal/text contexts and chunked-prefill scenarios.

## Problem

`merge_evs_mm_embeds` wrote each request's EVS-adjusted token IDs at a fixed offset 0:

```python
input_ids[: context_ids.shape[0]] = context_ids
```

This assumes multimodal context requests are a contiguous prefix of the flattened `input_ids`. With in-flight batching that's false - any text-only context request scheduled ahead of the EVS one gets silently clobbered (independent of chunked prefill). A guard (`_validate_evs_context_batch`) merely *rejected* such batches as a stopgap.

```
batch layout:  [ A text-only ] [ B EVS video ] [ decode ]
write target:  input_ids[0 : len(B)]
result:        B's IDs land on top of A  (A clobbered)
```

## Fix

Record where each request begins in the flattened `input_ids` (known in the model engine as `len(input_ids)` before the buffer is extended) on a new `MultimodalParams.input_ids_start_offset`, and write each request into its own span:

```python
for multimodal_param, context_part in zip(multimodal_params, context_parts, strict=True):
    start = multimodal_param.input_ids_start_offset
    input_ids[start : start + context_part.shape[0]] = context_part
```

Safe because the merge is **length-preserving**: the slot is pre-sized to the post-EVS token count and EVS only redistributes tokens across tubelets, never growing the span. The `_validate_evs_context_batch` guard is removed.

```
batch layout:  [ A text-only ] [ B EVS video ] [ decode ]
write target:  input_ids[len(A) : len(A) + len(B)]
result:        only B's span written  (A and decode untouched)
```

## Changes

- `inputs/multimodal.py` - add `input_ids_start_offset: int = 0` to `MultimodalParams`.
- `_torch/pyexecutor/model_engine.py` - record the offset in both context-prep paths.
- `_torch/models/modeling_nemotron_nano.py` - per-request offset writes; drop `_validate_evs_context_batch`.

## Test Coverage

`test_nemotron_nano_preprocessing.py::TestMergeEvsMMEmbeds` - replaced the rejection test with positive coverage: text-only before/after, interspersed multi-request, and chunked prefill with a preceding text-only request.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

### CODE REVIEW

**Comment by moraxu** [comment]
@coderabbitai full review

File: tensorrt_llm/_torch/pyexecutor/model_engine.py [hunk 1]
```diff
@@ -2677,7 +2681,8 @@ def _prepare_tp_inputs(
 
             multimodal_params = MultimodalParams(
                 multimodal_data=request.py_multimodal_data,
-                multimodal_runtime=py_multimodal_runtime)
+                multimodal_runtime=py_multimodal_runtime,
+                input_ids_start_offset=context_start_idx)
```
Comment (2ez4bz) [inline tensorrt_llm/_torch/pyexecutor/model_engine.py:2739]: Sorry, I may be missing something - how is this an offset? Aren't we always counting from zero when we define `context_start_idx = len(input_ids)` at line 2640?

File: tests/unittest/_torch/modeling/test_nemotron_nano_preprocessing.py [hunk 2]
```diff
@@ -1124,6 +1119,142 @@ def test_chunked_prefill_first_chunk_can_be_shorter_than_full_context(self):
...
+                decode.clone(),
+            ]
+        )
+
+        result = NemotronH_Nano_VL_V2.merge_evs_mm_embeds(
+            model, num_tokens_in_videos, [b_param, d_param], input_ids
+        )
+
+        expected_b = torch.tensor([_IMG_START] + [_IMG_CTX_ID] * 3 + [_IMG_END], dtype=torch.long)
+        assert (result[0:3] == a_text).all()
```
Comment (2ez4bz) [inline tests/unittest/_torch/modeling/test_nemotron_nano_preprocessing.py:1200]: Nit: could just construct the entire expected tensor and have a single `assert (x == y).all()?

File: tensorrt_llm/_torch/pyexecutor/model_engine.py (hunk 1 above)
Comment (moraxu) [reply tensorrt_llm/_torch/pyexecutor/model_engine.py:2739]: So `input_ids` is initialized once, outside the loop at line 2595 (`input_ids = []`), then this loop accumulates every request's tokens into that same list via `input_ids.extend(prompt_tokens)` above. So `context_start_idx = len(input_ids)` is the running length so far - it's `0` only on the first iteration.

File: tests/unittest/_torch/modeling/test_nemotron_nano_preprocessing.py (hunk 2 above)
Comment (moraxu) [reply tests/unittest/_torch/modeling/test_nemotron_nano_preprocessing.py:1200]: I think I will keep it as is - those per-span asserts are verifying which spans get written vs. left untouched (the clobbering regression). We'd have to do more digging which region broke exactly with just `assert (x == y).all()`

File: tensorrt_llm/_torch/pyexecutor/model_engine.py (hunk 1 above)
Comment (2ez4bz) [reply tensorrt_llm/_torch/pyexecutor/model_engine.py:2739]: Oh I see. Thanks for explaining.

**Review by 2ez4bz — APPROVED** [review]

File: tensorrt_llm/inputs/multimodal.py [hunk 3]
```diff
@@ -496,6 +502,7 @@ class MultimodalParams:
     multimodal_input: Optional[MultimodalInput] = None
     multimodal_data: Optional[Dict[str, Any]] = field(default_factory=dict)
     multimodal_runtime: Optional[MultimodalRuntimeData] = None
+    input_ids_start_offset: int = 0
```
Comment (achartier) [inline tensorrt_llm/inputs/multimodal.py:505]: Is the expectation that the user will always override this value? If so, maybe it should be an Optional and add a check for None?

File: tensorrt_llm/_torch/models/modeling_nemotron_nano.py [hunk 4]
```diff
@@ -2882,24 +2847,29 @@ def merge_evs_mm_embeds(
...
+        # contiguous prefix: text-only context requests carry no multimodal_params
+        # (so they never appear here) yet still occupy input_ids, and they may
+        # precede or sit between multimodal ones. Each write therefore starts at
+        # the request's recorded flattened offset. The merge is length-preserving
+        # (the slot was pre-sized to the post-EVS retained count), so every span
+        # ends exactly where the next request begins and `input_ids` is never
+        # grown. Under chunked prefill, multimodal_runtime narrows each context
+        # part to the current chunk so decode ids after it are preserved.
+        for multimodal_param, context_part in zip(multimodal_params, context_parts, strict=True):
+            start = multimodal_param.input_ids_start_offset
```
Comment (achartier) [inline tensorrt_llm/_torch/models/modeling_nemotron_nano.py:2862]: Maybe add a check that it's >= 0

File: tensorrt_llm/_torch/models/modeling_nemotron_nano.py [hunk 5]
```diff
@@ -2882,24 +2847,29 @@ def merge_evs_mm_embeds(
...
+            end = start + context_part.shape[0]
```
Comment (achartier) [inline tensorrt_llm/_torch/models/modeling_nemotron_nano.py:2863]: Should we also check that start is before end?

File: tensorrt_llm/_torch/models/modeling_nemotron_nano.py (hunk 4 above)
Comment (moraxu) [reply tensorrt_llm/_torch/models/modeling_nemotron_nano.py:2862]: Added in https://github.com/NVIDIA/TensorRT-LLM/pull/15016/commits/3728c57df0d4a05c80231290ef607abd4b45589a

File: tensorrt_llm/_torch/models/modeling_nemotron_nano.py (hunk 5 above)
Comment (moraxu) [reply tensorrt_llm/_torch/models/modeling_nemotron_nano.py:2863]: Added in https://github.com/NVIDIA/TensorRT-LLM/pull/15016/commits/3728c57df0d4a05c80231290ef607abd4b45589a

File: tensorrt_llm/inputs/multimodal.py (hunk 3 above)
Comment (moraxu) [reply tensorrt_llm/inputs/multimodal.py:505]: This is engine-populated and the first request in a batch legitimately starts at offset `0`

**Review by achartier — APPROVED** [review]

**Comment by 2ez4bz** [comment]
first try
